src/agent/rustjail/src/cgroups/fs/mod.rs

```diff
@@ -27,1120 +43,848 @@ macro_rules! sl {
-        "cpuacct".to_string()
-    }
+    fn freeze(&self, state: FreezerState) -> Result<()> {
+        let h = cgroups::hierarchies::auto();
+        let h = Box::new(&*h);
+        let cg = load_or_create(h, &self.cpath);
+        let freezer_controller: &FreezerController = cg.controller_of().unwrap();
+        match state {
+            FreezerState::Thawed => {
+                freezer_controller.thaw();
+            }
+            FreezerState::Frozen => {
+                freezer_controller.freeze();
+            }
+            _ => {
+                return Err(nix::Error::Sys(Errno::EINVAL).into());
+            }
+        }
 
-    fn set(&self, _dir: &str, _r: &LinuxResources, _update: bool) -> Result<()> {
         Ok(())
     }
-}
 
-fn get_cpuacct_percpu_usage(dir: &str) -> Result<Vec<u64>> {
-    let mut m = Vec::new();
-    let file = format!("{}/{}", dir, CPUACCT_PERCPU);
-
-    for n in fs::read_to_string(file.as_str())?.split(' ') {
-        m.push(n.trim().parse::<u64>()?);
+    fn destroy(&mut self) -> Result<()> {
+        let h = cgroups::hierarchies::auto();
+        let h = Box::new(&*h);
+        let cg = load(h, &self.cpath);
+        if cg.is_some() {
+            cg.unwrap().delete();
+        }
+        Ok(())
     }
 
-    Ok(m)
-}
-
-fn get_percpu_usage(dir: &str, file: &str) -> Result<Vec<u64>> {
-    let mut m = Vec::new();
-    let p = format!("{}/{}", dir, file);
-    info!(sl!(), "{}", p.as_str());
+    fn get_pids(&self) -> Result<Vec<pid_t>> {
+        let h = cgroups::hierarchies::auto();
+        let h = Box::new(&*h);
+        let cg = load_or_create(h, &self.cpath);
+        let mem_controller: &MemController = cg.controller_of().unwrap();
+        let pids = mem_controller.tasks();
+        let result = pids.iter().map(|x| x.pid as i32).collect::<Vec<i32>>();
 
-    for n in fs::read_to_string(p.as_str())?.split(' ') {
-        info!(sl!(), "{}", n);
-        if !n.trim().is_empty() {
-            m.push(n.trim().parse::<u64>()?);
-        }
+        Ok(result)
     }
-
-    Ok(m)
 }
 
-impl CpuAcct {
-    fn get_stats(&self, dir: &str) -> Result<CpuUsage> {
-        let h = get_param_key_u64(dir, CPUACCT_STAT)?;
-
-        let usage_in_usermode =
-            (((*h.get("user").unwrap() * NANO_PER_SECOND) as f64) / *CLOCK_TICKS) as u64;
-        let usage_in_kernelmode =
-            (((*h.get("system").unwrap() * NANO_PER_SECOND) as f64) / *CLOCK_TICKS) as u64;
-
-        info!(sl!(), "stat");
-
-        let total_usage = get_param_u64(dir, CPUACCT_USAGE)?;
-        info!(sl!(), "usage");
-
-        let percpu_usage = get_percpu_usage(dir, CPUACCT_PERCPU)?;
-        info!(sl!(), "percpu");
-
-        Ok(CpuUsage {
-            total_usage,
-            percpu_usage,
-            usage_in_kernelmode,
-            usage_in_usermode,
-            unknown_fields: UnknownFields::default(),
-            cached_size: CachedSize::default(),
-        })
+fn string_to_device_type(s: &String) -> DeviceType {
```

Comment:
This function does not feel like it belongs to this module.

Reply:
Yes, I found the method from `cgroup-rs` package.